File: taewan/08.16/17.cpp
#include<iostream>
#include<string>
#include<vector>
using namespace std;

class Solution {
public:
    vector<string> answer;    
    
    void recursive(string str, string digits, int depth) {
        if(depth == digits.length()) {      //문자열의 끝에 도달하면 answer에 push해주기
            if(str != "") answer.push_back(str);
            return ;
        }
        switch(digits[depth]) {     //한 단어씩 더하면서 재귀를 돌린다.
            case '2':
                recursive(str + 'a', digits, depth + 1);
                recursive(str + 'b', digits, depth + 1);
                recursive(str + 'c', digits, depth + 1);                
                break;
            case '3':
                recursive(str + 'd', digits, depth + 1);
                recursive(str + 'e', digits, depth + 1);
                recursive(str + 'f', digits, depth + 1);
                break;
            case '4':
                recursive(str + 'g', digits, depth + 1);
                recursive(str + 'h', digits, depth + 1);
                recursive(str + 'i', digits, depth + 1);
                break;
            case '5':
                recursive(str + 'j', digits, depth + 1);
                recursive(str + 'k', digits, depth + 1);
                recursive(str + 'l', digits, depth + 1);                
                break;
            case '6':
                recursive(str + 'm', digits, depth + 1);
                recursive(str + 'n', digits, depth + 1);
                recursive(str + 'o', digits, depth + 1);                
                break;
            case '7':
                recursive(str + 'p', digits, depth + 1);
                recursive(str + 'q', digits, depth + 1);
                recursive(str + 'r', digits, depth + 1);
                recursive(str + 's', digits, depth + 1);
                break;
            case '8':
                recursive(str + 't', digits, depth + 1);
                recursive(str + 'u', digits, depth + 1);
                recursive(str + 'v', digits, depth + 1);
                break;
            case '9':
                recursive(str + 'w', digits, depth + 1);
                recursive(str + 'x', digits, depth + 1);
                recursive(str + 'y', digits, depth + 1);
                recursive(str + 'z', digits, depth + 1);
                break;
        }
        return ;
    }
    
    vector<string> letterCombinations(string digits) {
        recursive("", digits, 0);
        return answer;
    }
};

int main(void){
  return 0;
}

File: deprecated/hyeyoung/17.py
class Solution:
    def letterCombinations(self, digits: str) -> List[str]:

        numToWord = {'2': ['a', 'b', 'c'],
                     '3': ['d', 'e', 'f'],
                     '4': ['g', 'h', 'i'],
                     '5': ['j', 'k', 'l'],
                     '6': ['m', 'n', 'o'],
                     '7': ['p', 'q', 'r', 's'],
                     '8': ['t', 'u', 'v'],
                     '9': ['w', 'x', 'y', 'z']}

        def backTracking(word, digit):
            if len(digit) == 0:
                answer.append(word)
            else:
                for alpa in numToWord[digit[0]]:
                    backTracking(word + alpa, digit[1:])  # 앞글자를 땐 나머지도 반복연산한다

        answer = []
        if digits:
            backTracking("", digits)
        return answer